### ISSUE TITLE
SBT doesn't cache submodule compilation results from run+cancel

### ISSUE THREAD

**Issue by jdrphillips** [sbt/sbt#5585]
When invoking `SBT run` or `sbt> run` **and the execution is cancelled**, if there is incremental compilation that needed performing in  a submodule, SBT does not cache these results. The same incremental compilation takes place every time you call `run` and cancel execution.

For example, running a server or a polling service.

### Expectation

SBT should record the results of all compilations.

### Versions

Observed in `1.3.11` and `1.3.9`

### Reproduction Steps

Clone this repo: https://github.com/jdrphillips/sbt-testing-2

Perform the following steps:

1. `sbt main/compile`
2. Mutate a file in module `base` in any way
3. `sbt main/run`
4. Cancel execution once program is running
5. `sbt main/run`
6. Cancel execution once program is running

Observe that the `base/compile` phase is run in both steps 3 and 5. It should only need be run in step 3.

**Comment by eed3si9n** [comment]
Thanks for the report. Could you provide reproduction steps so we can reproduce your problem on _our computers_?

**Comment by jdrphillips** [comment]
You're quite right, sorry. The real scenario was more nuanced than I thought when I came to produce a test repo for you.

I have updated with the info required to reproduce and clarify the scenario

**Comment by eed3si9n** [comment]
Cool. Thanks!

**Comment by eatkins** [comment]
I have noticed this behavior myself. I think it also happens to subprojects if you cancel while compiling a dependent project. Without looking at any code, my guess is that the fix would be in `EvaluateTask.scala`. There are probably task completion callbacks that aren't invoked if any task fails. In my opinion, this is a bug in task cancellation that is more general than just an incremental compilation bug.

**Comment by eed3si9n** [comment]
Likely fixed in sbt 2.x.
